[SPARK-26946][SQL] Identifiers for multi-catalog

jzhuge opened the pull request:
## What changes were proposed in this pull request?

- Support N-part identifier in SQL
- N-part identifier extractor in Analyzer

## How was this patch tested?

- A new unit test suite ResolveMultipartRelationSuite
- CatalogLoadingSuite

@rblue @cloud-fan @mccheah

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

dongjoon-hyun commented:
ok to test.

SparkQA commented:
**[Test build #102680 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102680/testReport)** for PR 23848 at commit [`86b9a4f`](https://github.com/apache/spark/commit/86b9a4f29fce6757a3760f5d89f89d6920e810c9).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8218/
Test PASSed.

SparkQA commented:
**[Test build #102680 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102680/testReport)** for PR 23848 at commit [`86b9a4f`](https://github.com/apache/spark/commit/86b9a4f29fce6757a3760f5d89f89d6920e810c9).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class UnresolvedRelation(table: CatalogTableIdentifier)`
  * `sealed trait IdentifierWithOptionalDatabaseAndCatalog `
  * `case class CatalogTableIdentifier(table: String, database: Option[String], catalog: Option[String])`
  * `class TableIdentifier(table: String, db: Option[String])`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102680/
Test PASSed.

jzhuge committed:
[SPARK-26946][SQL] Identifiers for multi-catalog

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8678/
Test PASSed.

SparkQA commented:
**[Test build #103232 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103232/testReport)** for PR 23848 at commit [`372e38f`](https://github.com/apache/spark/commit/372e38faa52a411c8e3c8f35d5286bd4ca576dfd).

SparkQA commented:
**[Test build #103232 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103232/testReport)** for PR 23848 at commit [`372e38f`](https://github.com/apache/spark/commit/372e38faa52a411c8e3c8f35d5286bd4ca576dfd).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public class CatalogNotFoundException extends RuntimeException `
  * `case class CatalogIdentifier(space: CatalogIdentifier.Namespace, name: String)`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103232/
Test FAILed.

jzhuge commented:
Looking at the build failure

jzhuge committed:
Fix build failure

jzhuge committed:
Change CatalogRef to return CatalogPlugin and CatalogIdentifier

SparkQA commented:
**[Test build #103234 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103234/testReport)** for PR 23848 at commit [`c434f4c`](https://github.com/apache/spark/commit/c434f4c957d8c9d08a489eaa03955ae7b544e92e).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8680/
Test PASSed.

SparkQA commented:
**[Test build #103234 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103234/testReport)** for PR 23848 at commit [`c434f4c`](https://github.com/apache/spark/commit/c434f4c957d8c9d08a489eaa03955ae7b544e92e).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103234/
Test FAILed.

jzhuge committed:
Fix test failure in CatalogLoadingSuite

SparkQA commented:
**[Test build #103255 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103255/testReport)** for PR 23848 at commit [`feed4a3`](https://github.com/apache/spark/commit/feed4a355a35eda1b97b87068eeabb5888b28866).

SparkQA commented:
**[Test build #103255 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103255/testReport)** for PR 23848 at commit [`feed4a3`](https://github.com/apache/spark/commit/feed4a355a35eda1b97b87068eeabb5888b28866).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103255/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8701/
Test PASSed.

cloud-fan commented on sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4:
```diff
@@ -63,6 +63,10 @@ singleTableIdentifier
     : tableIdentifier EOF
     ;
 
+singleMultiPartIdentifier
```
This is a top-level parser entry used in `ParserInterface`. I don't think we need it now for catalog identifier.

jzhuge replied:
True, only my test case uses it to parse a table name into a sequence. I will remove it.

rdblue replied:
Won't we need this eventually for parsing names passed to `saveAsTable`? Why not add it now?

jzhuge replied:
When I start to to convert SELECT, INSERT, and DROP code path to support multi-catalog, this parse function is needed, e.g,
```
  override def visitTable(ctx: TableContext): LogicalPlan = withOrigin(ctx) {
    UnresolvedIdentifier(visitMultiPartIdentifier(ctx.multiPartIdentifier))
  }

 override def visitTableName(ctx: TableNameContext): LogicalPlan = withOrigin(ctx) {
    val tableId = visitMultiPartIdentifier(ctx.multiPartIdentifier())
    val table = mayApplyAliasPlan(ctx.tableAlias, UnresolvedIdentifier(tableId))
    table.optionalMap(ctx.sample)(withSample)
  }
```

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/identifiers.scala:
```diff
@@ -105,3 +105,15 @@ case class FunctionIdentifier(funcName: String, database: Option[String])
+ * @param space Namespace in the catalog
+ * @param name Object name
+ */
+case class CatalogIdentifier(space: CatalogIdentifier.Namespace, name: String)
```
I always fail to understand the name `CatalogIdentifier`.

It's an identifier of a table in a certain catalog, instead of an identifier of a catalog. Basically it's still a table identifier.

jzhuge replied:
Yes, it is similar to table identifier.
How about renaming `CatalogIdentifier` to `ObectIdentifier`?
And redefines `CatalogIdentifier` as:
```
CatalogIdentifier(catalog: Option[String], ident: ObectIdentifier)
```

jzhuge replied:
Unfortunately `ObjectIdentifier` is not ideal either. Any suggestion?
Otherwise we may live with `CatalogIdentifier` that identifies something inside a catalog.

rdblue replied:
I think `CatalogIdentifier` is fine, unless someone can think of a better option.

rdblue replied:
Looks like `Identifier` isn't used. If this is in the v2 namespace, it would be fine to simply call the interface `Identifier`.

jzhuge committed:
Add constructor CatalogIdentifier(name)

rdblue commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/catalog/v2/ResolveMultipartRelationSuite.scala:
```diff
@@ -0,0 +1,98 @@
+
+  private val analyzer = makeAnalyzer(caseSensitive = false)
+
+  private val catalogs = Seq("prod", "test").map(name => name -> new TestCatalogPlugin(name)).toMap
```
Why not use the plugin system to instantiate these catalogs?

mccheah replied:
It depends on if this test should also be exercising the code path that does said loading, or if we want a separate unit test to verify that behavior in isolation. I'm usually a proponent of each test covering a specific feature and that feature only, so as long as we have another unit test that exercises plugin loading then I think this is fine.

mccheah replied:
To be clear I think we should be avoiding plugin loading in these tests.

rdblue replied:
+1, good call.

mccheah commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -95,13 +96,40 @@ object AnalysisContext {
+  }
+
+  object CatalogRef {
+    def unapply(parts: Seq[String]): Option[(Option[CatalogPlugin], CatalogIdentifier)] =
```
Nit: The nesting in the return type is a little tricky to follow, particularly `Option[(Option[...])...]`. Is it worth wrapping `(Option[CatalogPlugin], CatalogIdentifier])` in its own dedicated case class? Then it's `Option[CatalogThing]` (I actually don't know what a good name for such a case class is...) which reads nicer.

jzhuge replied:
Same here, for lack of a good name.`CatalogRef` is not ideal either. I will put some thoughts in the next few days. Suggestions are welcome!

jzhuge replied:
Fixed

mccheah commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/catalog/v2/ResolveMultipartRelationSuite.scala:
```diff
@@ -0,0 +1,98 @@
+private case class TestMultipartAnalysis(analyzer: Analyzer) extends Rule[LogicalPlan] {
+
+  override def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperators  {
+
```
Nit: Don't think we want the whitespace here.

jzhuge replied:
Fixed the 2 white spaces after resolveOperators

mccheah commented:
Sorry for breaking up my review into individual comments. I think this looks ok short of some style changes.

rdblue commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/identifiers.scala:
```diff
@@ -105,3 +105,15 @@ case class FunctionIdentifier(funcName: String, database: Option[String])
+ * @param space Namespace in the catalog
+ * @param name Object name
+ */
+case class CatalogIdentifier(space: CatalogIdentifier.Namespace, name: String)
```
I think that these should be in the `org.apache.spark.sql.catalyst.catalog.v2` namespace. That separates the new identifiers from the old ones and keeps them with the other v2 additions.

Also, we will need to translate this to Java. I wrote the SPIP using scala conventions for clarity, but we should use types that are Java-friendly. For example, Java implementations should not be passed `Seq` so namespace should be `Array[String]` instead.

rdblue commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/catalog/v2/ResolveMultipartRelationSuite.scala:
```diff
@@ -0,0 +1,98 @@
+    val conf = new SQLConf().copy(SQLConf.CASE_SENSITIVE -> caseSensitive)
+    new Analyzer(Some(lookupCatalog _), null, conf) {
+      override val extendedResolutionRules =
+        EliminateSubqueryAliases :: TestMultipartAnalysis(this) :: Nil
```
This uses a lot of temporary classes to simulate future rules that match multi-part identifiers. I think I would rather include an update that adds new `UnresolvedRelation` nodes and uses them instead of test plan nodes, but I'd be interested to hear whether @cloud-fan agrees.

jzhuge replied:
OK either way. I have already converted SELECT/INSERT/DROP code paths to support multi-catalog in my private 2.3 branch. Pretty straightforward. Converting CREATE would be a lot easier with Ryan's PR 24029.

rdblue commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -95,13 +96,40 @@ object AnalysisContext {
+  }
+
+  object CatalogRef {
+    def unapply(parts: Seq[String]): Option[(Option[CatalogPlugin], CatalogIdentifier)] =
```
I like this. It provides a good way to add support in match expressions. You might consider adding one that translates to `TableIdentifier` as well:

```scala
object V1Identifier {
  def unapply(parts: Seq[String]): Option[TableIdentifier] = ...
}
```

That would match if there are only 2 identifier parts and the first is not a known catalog name.

mccheah replied:
To clarify because Github isn't putting your comment in a thread, is this in response to https://github.com/apache/spark/pull/23848#discussion_r266179932? If so, is "I like this" meaning "I like the case class" that above comment has suggested, or, that you prefer the way it is currently?

Just trying to disambiguate.

rdblue replied:
This wasn't part of the other thread. I like using an extractor as a helper for writing rules. This will help us produce rules that are cleaner.

jzhuge commented on sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4:
```diff
@@ -554,6 +558,10 @@ rowFormat
       (NULL DEFINED AS nullDefinedAs=STRING)?                                       #rowFormatDelimited
     ;
 
+multiPartIdentifier
```
Any name better than `multiPartIdentifier`?

rdblue replied:
I think this is good. It is descriptive.

jzhuge committed:
Review feedbacks

Create org.apache.spark.sql.catalog.v2.Identifier and IdentifierImpl.
Inherit CatalogIdentifier from v2.Identifier.
Encapsulate lookupCatalog and extractor into trait LookupCatalog.
SqlBase.g4: Replace MultiPart with Multipart.
Rename and simplify the unit test ResolveMultipartIdentifierSuite.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/9024/
Test PASSed.

SparkQA commented:
**[Test build #103623 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103623/testReport)** for PR 23848 at commit [`54280ed`](https://github.com/apache/spark/commit/54280ed2c03fb2ce55eac4bb02d2765782113551).

SparkQA commented:
**[Test build #103623 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103623/testReport)** for PR 23848 at commit [`54280ed`](https://github.com/apache/spark/commit/54280ed2c03fb2ce55eac4bb02d2765782113551).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public class IdentifierImpl implements Identifier `
  * `trait LookupCatalog `
  * `case class CatalogIdentifier(namespace: Array[String], name: String) extends Identifier`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103623/
Test FAILed.

rdblue commented on sql/catalyst/src/main/java/org/apache/spark/sql/catalog/v2/IdentifierImpl.java:
```diff
@@ -0,0 +1,43 @@
+
+  @Override
+  public String[] namespace() {
+    return new String[0];
```
Looks like `namespace` and `name` weren't implemented.

jzhuge replied:
Fixed

rdblue commented on sql/catalyst/src/main/java/org/apache/spark/sql/catalog/v2/Identifier.java:
```diff
@@ -0,0 +1,37 @@
+/**
+ * Identifies an object in a catalog.
+ */
+@Evolving
```
I think this should be experimental.

jzhuge replied:
+1

rdblue commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/identifiers.scala:
```diff
@@ -105,3 +107,11 @@ case class FunctionIdentifier(funcName: String, database: Option[String])
+ * @param namespace the namespace in the catalog
+ * @param name the object name
+ */
+case class CatalogIdentifier(namespace: Array[String], name: String) extends Identifier
```
I don't think this is needed any more.

jzhuge replied:
+1

rdblue commented on sql/catalyst/src/main/java/org/apache/spark/sql/catalog/v2/IdentifierImpl.java:
```diff
@@ -0,0 +1,43 @@
+
+package org.apache.spark.sql.catalog.v2;
+
+public class IdentifierImpl implements Identifier {
```
Does this need to be public? I think `of` could be a static method on `Identifier` and then this implementation could be package-private.

jzhuge replied:
+1

rdblue commented:
@jzhuge, this looks really close to being ready to me!

jzhuge committed:
More review feedbacks

Add extractor LookupCatalog.AsTableIdentifier and a unit test.
Remove CatalogIdentifier.

SparkQA commented:
**[Test build #103654 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103654/testReport)** for PR 23848 at commit [`5e18fcf`](https://github.com/apache/spark/commit/5e18fcf427e74698f65052a9cde92e2adda6d791).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/9054/
Test PASSed.

SparkQA commented:
**[Test build #103654 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103654/testReport)** for PR 23848 at commit [`5e18fcf`](https://github.com/apache/spark/commit/5e18fcf427e74698f65052a9cde92e2adda6d791).
 * This patch **fails Java style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class IdentifierImpl implements Identifier `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103654/
Test FAILed.

jzhuge committed:
Fix checkstyle error

SparkQA commented:
**[Test build #103658 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103658/testReport)** for PR 23848 at commit [`ce96f84`](https://github.com/apache/spark/commit/ce96f842d854d31060e7dbe13e7f88b4118584df).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/9057/
Test PASSed.

SparkQA commented:
**[Test build #103658 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103658/testReport)** for PR 23848 at commit [`ce96f84`](https://github.com/apache/spark/commit/ce96f842d854d31060e7dbe13e7f88b4118584df).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103658/
Test FAILed.

dilipbiswal commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/9064/
Test PASSed.

SparkQA commented:
**[Test build #103666 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103666/testReport)** for PR 23848 at commit [`ce96f84`](https://github.com/apache/spark/commit/ce96f842d854d31060e7dbe13e7f88b4118584df).

SparkQA commented:
**[Test build #103666 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103666/testReport)** for PR 23848 at commit [`ce96f84`](https://github.com/apache/spark/commit/ce96f842d854d31060e7dbe13e7f88b4118584df).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103666/
Test PASSed.

rdblue commented on sql/catalyst/src/main/java/org/apache/spark/sql/catalog/v2/IdentifierImpl.java:
```diff
@@ -0,0 +1,45 @@
+import org.apache.spark.annotation.Experimental;
+
+/**
+ *  An [[Identifier]] implementation.
```
Minor: Looks like Scaladoc conventions used in Javadoc. This should be `{@link Identifier}`.

rdblue commented:
+1

This looks good to me. @cloud-fan, do you have any more review comments?

jzhuge committed:
Fix javadoc for IdentifierImpl

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/9082/
Test PASSed.

SparkQA commented:
**[Test build #103685 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103685/testReport)** for PR 23848 at commit [`11c7871`](https://github.com/apache/spark/commit/11c7871de02359f4f415ca1f587614e0b289cd20).

cloud-fan commented on sql/catalyst/src/main/java/org/apache/spark/sql/catalog/v2/Identifier.java:
```diff
@@ -0,0 +1,41 @@
+ * Identifies an object in a catalog.
+ */
+@Experimental
+public interface Identifier {
```
Shall we use a class directly? I don't see much value of using an interface here, as it has only one implementation.

rdblue replied:
This allows us more flexibility than a single concrete class. Changing a class to an interface is not a binary compatible change, so using an interface is the right thing to do.

cloud-fan replied:
Then I suggest we move the impl class to a private package like `org.apache.spark.sql.catalyst`. Also the static method should be moved to the impl class as well, as we only create it inside Spark.

rdblue replied:
The implementation class is package-private. If we were to move it to a different package, we would need to make it public for the `of` factory method, which would increase its visibility, not decrease it.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalog/v2/LookupCatalog.scala:
```diff
@@ -0,0 +1,62 @@
+import org.apache.spark.sql.catalyst.TableIdentifier
+
+@Experimental
+trait LookupCatalog {
```
Why it's a trait?

My understanding is this PR adds the class of the catalog object identifier and the related parser support. I don't think we have a detailed design of how analyzer looks up catalog yet.

rdblue replied:
This trait provides extractors, similar to a trait like `PredicateHelper`. These implement the [resolution rules from the SPIP](https://docs.google.com/document/d/1jEcvomPiTc5GtB9F7d2RTVVpMY64Qy7INCA_rFEd9HQ/edit#heading=h.dwrdqt8xe2wu) using a generic catalog lookup provided by the implementation.

This decouples the resolution rules from how the analyzer looks up catalogs and provides convenient extractors that implement those rules.

cloud-fan replied:
then this should be an internal trait under a private package like `org.apache.spark.sql.catalyst`

RussellSpitzer commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalog/v2/LookupCatalog.scala:
```diff
@@ -0,0 +1,62 @@
+    }
+  }
+
+  object AsTableIdentifier {
```
We may want to highlight that this is an extractor for supporting old legacy use-cases, modern rules and such should use CatalogObjectIdentifier and assume multipart extraction? Just a comment on the code for future dev's using these extractors and wondering which they should implement.

jzhuge replied:
Sounds good

jzhuge committed:
Add test cases for back-quoted identifiers

Add comment for AsTableIdentifier to emphasize legacy support only.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/9129/
Test PASSed.

SparkQA commented:
**[Test build #103750 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103750/testReport)** for PR 23848 at commit [`3bb4485`](https://github.com/apache/spark/commit/3bb448503f35e9e035480401261702bcc3114e7a).

SparkQA commented:
**[Test build #103750 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103750/testReport)** for PR 23848 at commit [`3bb4485`](https://github.com/apache/spark/commit/3bb448503f35e9e035480401261702bcc3114e7a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103750/
Test PASSed.

mccheah reviewed (approved):
@cloud-fan have we addressed all your comments, or did you have any other feedback you wanted to give? Would like to merge this soon to unblock other V2 work, particularly table catalogs.

Possibly try to merge before EOD Pacific time today, at the very latest before end of week?

For everyone else following, please feel free to leave any feedback we would like to address before this goes in.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -95,13 +96,19 @@ object AnalysisContext {
     this(catalog, conf, conf.optimizerMaxIterations)
   }
 
+  def this(lookupCatalog: Option[(String) => CatalogPlugin], catalog: SessionCatalog,
```
who will call this constructor? I feel we are adding too much code for future use only. Can we add them when they are needed? It will be good if this PR only add the identifier interface and impl class, and the related parser rules, which is pretty easy to justify.

rdblue replied:
@cloud-fan, I think this commit is reasonably self-contained. Nit-picking about whether a constructor is added in this commit or the next isn't adding much value.

Keep in mind that we make commits self-contained to decrease conflicts and increase the rate at which we can review and accept patches. Is putting this in the next commit really worth the time it takes to change and test that change, if it means that this work is delayed another day?

cloud-fan commented:
The parser part and identifier interface/impl class LGTM. The catalog lookup part looks reasonable but I'm not very confident without seeing the actual use case. To move things forward, I'm merging this. I may refactor this part after the table catalog gets it.

cloud-fan commented:
thanks, merging to master!

jzhuge commented:
Thanks @cloud-fan !

rdblue commented:
Thanks for merging, @cloud-fan, and thanks for working on this, @jzhuge!
